### PR TITLE
Add -gpl extension to FFmpeg (#419)

### PR DESCRIPTION
**Pull request by tmurakam**
- add '-nogpl' extension.
- cppbuild.sh: Disable GPL modules.
- setup-source.sh: patch postproc.java.

### CODE REVIEW

File: ffmpeg/cppbuild.sh [hunk 1]
```diff
@@ -10,9 +10,11 @@ fi
...
+if [[ "$EXTENSION" == *nogpl ]]; then
+    # minimal configuration to support MPEG-4 streams with H.264 and AAC as well as Motion JPEG
+    DISABLE="--disable-iconv --disable-libxcb --disable-opencl --disable-sdl2 --disable-bzlib --disable-lzma --disable-linux-perf --disable-everything"
+    ENABLE="--enable-shared --enable-runtime-cpudetect --enable-libopenh264 --enable-encoder=libopenh264 --enable-encoder=aac --enable-encoder=mjpeg --enable-decoder=h264 --enable-decoder=aac --enable-decoder=mjpeg --enable-parser=h264 --enable-parser=aac --enable-parser=mjpeg --enable-muxer=mp4 --enable-muxer=rtsp --enable-muxer=mjpeg --enable-demuxer=mov --enable-demuxer=rtsp --enable-demuxer=mjpeg --enable-protocol=file --enable-protocol=http --enable-protocol=rtp --enable-protocol=rtmp"
```
Comment (saudet) [inline ffmpeg/cppbuild.sh:16]: I think we can include more than that that isn't GPL? Maybe we should name this "minimal"?

Comment (saudet) [reply ffmpeg/cppbuild.sh:16]: Or maybe we could have the default build be "minimal", and move everything to the "-gpl" extension? That probably makes more sense.

Comment (tmurakam) [reply ffmpeg/cppbuild.sh:16]: Ok, I'll try to add -gpl extension and default to be minimal.

Comment (tmurakam) [reply ffmpeg/cppbuild.sh:16]: Also I'll check the minimal configuration.

**Comment by tmurakam** [comment]
I had updated following points:

* Add '-gpl' extension and remove '-nogpl' (Default is non-GPL)
* Add more options for default profile.

**Comment by saudet** [comment]
We only need to have one "extension" line like this in avutil and it should work:
https://github.com/bytedeco/javacpp-presets/blob/master/opencv/src/main/java/org/bytedeco/opencv/presets/opencv_core.java#L72
Can you let me know what happens with that?

We can also do it like this for modules that are only available with extension like postproc:
https://github.com/bytedeco/javacpp-presets/blob/master/opencv/src/main/java/org/bytedeco/opencv/presets/opencv_cudaarithm.java
That way, it doesn't get processed without extension.

**Comment by tmurakam** [comment]
> We only need to have one "extension" line like this in avutil and it should work:

I've tried it, but it does not work for default (without) extension build. (It works for -gpl extension).
When remove the "extension" line for default, the source code is not generated (src/gen folder is empty), and no shared library are included in the jar file.

> We can also do it like this for modules that are only available with extension like postproc:

It does not work, the compilation fails:

> [ERROR] /home/tmurakam/dev/javacpp-presets/ffmpeg/src/gen/java/org/bytedeco/ffmpeg/avfilter/AVFilterInOut.java:[18,1] package org.bytedeco.ffmpeg.postproc does not exist

**Comment by saudet** [comment]
It works for OpenCV, it'll work with FFmpeg for sure. Show me what you tried exactly.

**Comment by tmurakam** [comment]
> It works for OpenCV, it'll work with FFmpeg for sure. Show me what you tried exactly.

Here is my reproduce steps.

First clone my repo from https://github.com/javacpp-nogpl/javacpp-presets on CentOS 7 machine, then:

```
# checkout working branch. The branch has only have one extension for presets/*.java (except for postproc).
git checkout ffmpeg-nogpl-extension-work

# Build ffmpeg for default
mvn -pl ffmpeg install
```

The generated ffmpeg/target/ffmpeg-linux-x86_64.jar does not contains *.so files.

**Comment by saudet** [comment]
I don't see a `@Platform(extension = "-gpl")` line in avutil anywhere.

**Comment by tmurakam** [comment]
I understood how to write extension line.
I had updated the presets/*.java except for postprocess.

But I don't understand how to fix the postprocess.java, because the postprocess.h does not exist when --disable-gpl mode...

**Comment by saudet** [comment]
Check the CUDA module for OpenCV above. If there are includes given only
for an extension, they'll get ignored without extension.

**Comment by tmurakam** [comment]
> Check the CUDA module for OpenCV above. If there are includes given only for an extension, they'll get ignored without extension.

It seems it does not work for FFmpeg, because avfilter inherits postproces:

```
@Properties(
    inherit = {avformat.class, postproc.class, swresample.class, swscale.class},
```

Can we remove the inheritance?

**Comment by saudet** [comment]
That should still be fine.

**Comment by tmurakam** [comment]
Ok, I had fixed the postproc.

**Comment by saudet** [comment]
We should only need `@Platform(extension = "-gpl")` in avutil because it gets inherited.

**Comment by tmurakam** [comment]
I had removed unnecessary `@Platform(extension = "-gpl")` lines.

File: ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avfilter.java [hunk 2]
```diff
@@ -33,7 +33,7 @@
  * @author Samuel Audet
  */
 @Properties(
-    inherit = {avformat.class, postproc.class, swresample.class, swscale.class},
+    inherit = {avformat.class, /*postproc.class,*/ swresample.class, swscale.class},
```
Comment (saudet) [inline ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avfilter.java:36]: This doesn't work if we leave postproc?

Comment (tmurakam) [reply ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avfilter.java:36]: I'm trying to revert and test building...

Comment (tmurakam) [reply ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avfilter.java:36]: It works well, so I'll push the commit.

**Comment by tmurakam** [comment]
I had push the commit which revert the avfilter.java to inherit postproc.

File: ffmpeg/cppbuild.sh [hunk 3]
```diff
@@ -8,11 +8,17 @@ if [[ -z "$PLATFORM" ]]; then
...
+ENABLE="$ENABLE --enable-version3 --enable-zlib --enable-libmp3lame --enable-libspeex --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libvo-amrwbenc --enable-openssl --enable-libopenh264 --enable-libvpx --enable-libfreetype --enable-libopus"
+
+if [[ "$EXTENSION" == *gpl ]]; then
+    # Enable GPL and nonfree modules
+    ENABLE="$ENABLE --enable-gpl --enable-nonfree --enable-libx264 --enable-libx265"
+else
+    # Disabled GPL configuration to support MPEG-4 streams with H.264 and AAC as well as Motion JPEG
+    #DISABLE="$DISABLE --disable-libxcb --disable-everything"
+    ENABLE="$ENABLE --enable-libopenh264 --enable-encoder=libopenh264 --enable-encoder=aac --enable-encoder=mjpeg --enable-decoder=h264 --enable-decoder=aac --enable-decoder=mjpeg --enable-parser=h264 --enable-parser=aac --enable-parser=mjpeg --enable-muxer=mp4 --enable-muxer=rtsp --enable-muxer=mjpeg --enable-demuxer=mov --enable-demuxer=rtsp --enable-demuxer=mjpeg --enable-protocol=file --enable-protocol=http --enable-protocol=rtp --enable-protocol=rtmp"
+fi
```
Comment (saudet) [inline ffmpeg/cppbuild.sh:21]: Can we replace that with something like this for simplicity?

```java
DISABLE="--disable-iconv --disable-opencl --disable-sdl2 --disable-bzlib --disable-lzma --disable-linux-perf"
ENABLE="--enable-shared --enable-version3 --enable-runtime-cpudetect --enable-zlib --enable-libmp3lame --enable-libspeex --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libvo-amrwbenc --enable-openssl --enable-libopenh264 --enable-libvpx --enable-libfreetype --enable-libopus"

if [[ "$EXTENSION" == *gpl ]]; then
    # Enable GPL and nonfree modules
    ENABLE="$ENABLE --enable-gpl --enable-nonfree --enable-libx264 --enable-libx265"
fi

# minimal configuration to support MPEG-4 streams with H.264 and AAC as well as Motion JPEG
# DISABLE="--disable-iconv --disable-libxcb --disable-opencl --disable-sdl2 --disable-bzlib --disable-lzma --disable-linux-perf --disable-everything"
# ENABLE="--enable-shared --enable-runtime-cpudetect --enable-libopenh264 --enable-encoder=libopenh264 --enable-encoder=aac --enable-encoder=mjpeg --enable-decoder=h264 --enable-decoder=aac --enable-decoder=mjpeg --enable-parser=h264 --enable-parser=aac --enable-parser=mjpeg --enable-muxer=mp4 --enable-muxer=rtsp --enable-muxer=mjpeg --enable-demuxer=mov --enable-demuxer=rtsp --enable-demuxer=mjpeg --enable-protocol=file --enable-protocol=http --enable-protocol=rtp --enable-protocol=rtmp"
```

File: ffmpeg/platform/gpl/pom.xml [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+//                      requires static org.bytedeco.${javacpp.moduleId}.android.x86;
+//                      requires static org.bytedeco.${javacpp.moduleId}.android.x86_64;
+                      requires static org.bytedeco.${javacpp.moduleId}.linux.x86;
+                      requires static org.bytedeco.${javacpp.moduleId}.linux.x86_64;
+                      requires static org.bytedeco.${javacpp.moduleId}.linux.armhf;
+                      requires static org.bytedeco.${javacpp.moduleId}.linux.arm64;
+                      requires static org.bytedeco.${javacpp.moduleId}.linux.ppc64le;
+                      requires static org.bytedeco.${javacpp.moduleId}.macosx.x86_64;
+                      requires static org.bytedeco.${javacpp.moduleId}.windows.x86;
+                      requires static org.bytedeco.${javacpp.moduleId}.windows.x86_64;
```
Comment (saudet) [inline ffmpeg/platform/gpl/pom.xml:174]: The module names here are missing a `.gpl` suffix.

File: ffmpeg/pom.xml [hunk 5]
```diff
@@ -14,6 +14,10 @@
   <version>4.3.1-${project.parent.version}</version>
   <name>JavaCPP Presets for FFmpeg</name>
 
+  <properties>
+    <!-- <javacpp.platform.extension>-gpl</javacpp.platform.extension> -->
+  </properties>
+
```
Comment (saudet) [inline ffmpeg/pom.xml:20]: We can remove that.

File: ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java [hunk 6]
```diff
@@ -38,8 +38,8 @@
     global = "org.bytedeco.ffmpeg.global.avcodec",
     value = {
         @Platform(cinclude = {"<libavcodec/codec_id.h>", "<libavcodec/codec_desc.h>", "<libavcodec/codec_par.h>", "<libavcodec/packet.h>",
-                              "<libavcodec/bsf.h>", "<libavcodec/codec.h>", "<libavcodec/avcodec.h>", "<libavcodec/jni.h>", "<libavcodec/avfft.h>"},
-                  link = "avcodec@.58"),
+            "<libavcodec/bsf.h>", "<libavcodec/codec.h>", "<libavcodec/avcodec.h>", "<libavcodec/jni.h>", "<libavcodec/avfft.h>"},
+            link = "avcodec@.58"),
```
Comment (saudet) [inline ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java:42]: Please revert unnecessary formatting changes here.

**Review by saudet — CHANGES_REQUESTED** [review]
Let's also update the pom.xml file in README.md and the samples subdirectory with something like this:
```xml
<project>
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.bytedeco.ffmpeg</groupId>
    <artifactId>readfewframe</artifactId>
    <version>1.5.5-SNAPSHOT</version>
    <properties>
        <exec.mainClass>ReadFewFrame</exec.mainClass>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>ffmpeg-platform</artifactId>
            <version>4.3.1-1.5.5-SNAPSHOT</version>
        </dependency>

        <!-- Optional GPL build with (almost) everything enabled -->
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>ffmpeg-platform-gpl</artifactId>
            <version>4.3.1-1.5.5-SNAPSHOT</version>
        </dependency>
    </dependencies>
    <build>
        <sourceDirectory>.</sourceDirectory>
    </build>
</project>
```

File: ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java (hunk 6 above)
Comment (tmurakam) [reply ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java:42]: Ok.

File: ffmpeg/cppbuild.sh (hunk 3 above)
Comment (tmurakam) [reply ffmpeg/cppbuild.sh:21]: Hmm.... I think we can do it, but I think there are no "-enable-encoder/decoder/parser/protocol/muxer/demuxer" configuration in default (non-GPL) configuration. Is it okay?

Comment (saudet) [reply ffmpeg/cppbuild.sh:21]: Yes, the configuration is the same since they are not disabled explicitly.

File: ffmpeg/platform/gpl/pom.xml (hunk 4 above)
Comment (tmurakam) [reply ffmpeg/platform/gpl/pom.xml]: Fixed.

File: ffmpeg/pom.xml (hunk 5 above)
Comment (tmurakam) [reply ffmpeg/pom.xml]: Fixed.

File: ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java (hunk 6 above)
Comment (tmurakam) [reply ffmpeg/src/main/java/org/bytedeco/ffmpeg/presets/avcodec.java]: Fixed.

File: ffmpeg/cppbuild.sh (hunk 3 above)
Comment (tmurakam) [reply ffmpeg/cppbuild.sh:16]: Okay, I'll fix it as your comment.

**Comment by tmurakam** [comment]
I had updated the README.
I think "ffmpeg-platform" and "ffmpeg-platform-gpl" are exclusive, so we can use only one of them.
 
> ```
>         <dependency>
>             <groupId>org.bytedeco</groupId>
>             <artifactId>ffmpeg-platform</artifactId>
>             <version>4.3.1-1.5.5-SNAPSHOT</version>
>         </dependency>
> 
>         
>         <dependency>
>             <groupId>org.bytedeco</groupId>
>             <artifactId>ffmpeg-platform-gpl</artifactId>
>             <version>4.3.1-1.5.5-SNAPSHOT</version>
>         </dependency>
> ```

**Comment by saudet** [comment]
> I had updated the README.
> I think "ffmpeg-platform" and "ffmpeg-platform-gpl" are exclusive, so we can use only one of them.

No, they are not mutually exclusive. The extension can be given at runtime. Also update `sample/pom.xml`, thanks!

**Comment by tmurakam** [comment]
> No, they are not mutually exclusive. The extension can be given at runtime. Also update `sample/pom.xml`, thanks!

Ok, I understood. I fixed the README and samples/pom.xml.

File: ffmpeg/platform/gpl/pom.xml [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            <phase>package</phase>
+            <goals>
+              <goal>add-module-info</goal>
+            </goals>
+            <configuration>
+              <modules>
+                <module>
+                  <file>${project.build.directory}/${project.artifactId}.jar</file>
+                  <moduleInfoSource>
+                    module org.bytedeco.${javacpp.moduleId}.platform {
```
Comment (saudet) [inline ffmpeg/platform/gpl/pom.xml:162]: This one also needs a `.gpl` suffix.

**Review by saudet — CHANGES_REQUESTED** [review]

**Comment by saudet** [comment]
One last thing is that we'll need to add `-gpl` builds to `.travis.yml` and `.appveyor.yml` like this:
```xml
...
      - os: linux
        env: PROJ=ffmpeg OS=android-arm
        install: true
        script: ./ci/install-travis.sh
      - os: linux
        env: PROJ=ffmpeg OS=android-arm EXT=-gpl
        install: true
        script: ./ci/install-travis.sh
etc...

File: ffmpeg/platform/gpl/pom.xml (hunk 7 above)
Comment (tmurakam) [reply ffmpeg/platform/gpl/pom.xml]: Ok, fixed.

**Comment by tmurakam** [comment]
I had updated the .travis.yml and .appveyor.yml.

File: .appveyor.yml [hunk 8]
```diff
@@ -106,6 +106,12 @@ environment:
     - PROJ: "onnxruntime"
       OS: windows-x86_64
       EXT: -gpu
+    - PROJ: "ffmpeg"
+      OS: windows-x86
+      EXT: -gpl
+    - PROJ: "ffmpeg"
+      OS: windows-x86_64
+      EXT: -gpl
```
Comment (saudet) [inline .appveyor.yml:114]: Let's put those next to their default builds, like you did in `.travis.yml`.

**Review by saudet — CHANGES_REQUESTED** [review]

File: .appveyor.yml (hunk 8 above)
Comment (tmurakam) [reply .appveyor.yml]: Fixed.

**Review by saudet — APPROVED** [review]
